core/src/main/java/org/elasticsearch/index/mapper/BinaryRangeUtil.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.store.ByteArrayDataOutput;
+import org.apache.lucene.util.BytesRef;
+
+import java.io.IOException;
+import java.util.Set;
+
+class BinaryRangeUtil {
+
+    static BytesRef encodeLongRanges(Set<RangeFieldMapper.Range> ranges) throws IOException {
+        final byte[] encoded = new byte[5 + ((5 + 9) * 2) * ranges.size()];
+        ByteArrayDataOutput out = new ByteArrayDataOutput(encoded);
+        out.writeVInt(ranges.size());
+        for (RangeFieldMapper.Range range : ranges) {
+            byte[] encodedFrom = encode(((Number) range.from).longValue());
+            out.writeVInt(encodedFrom.length);
+            out.writeBytes(encodedFrom, encodedFrom.length);
+            byte[] encodedTo = encode(((Number) range.to).longValue());
+            out.writeVInt(encodedTo.length);
+            out.writeBytes(encodedTo, encodedTo.length);
+        }
+        return new BytesRef(encoded, 0, out.getPosition());
+    }
+
+    static byte[] encode(long l) {
+        int sign = 1; // means positive
+        if (l < 0) {
+            l = -1 - l;
+            sign = 0;
+        }
+        return encode(l, sign);
+    }
+
+    static byte[] encode(double d) {
+        long l;
+        int sign;
+        if (d < 0.0) {
+            l = Double.doubleToRawLongBits(-0d - d);
+            sign = 0;
+        } else {
+            l = Double.doubleToRawLongBits(d);
+            sign = 1; // means positive
+        }
+        return encode(l, sign);
```

Comment:
doubles should probably use a different encoding, but we can work on it as a follow-up